util/perftest.c

```diff
@@ -0,0 +1,116 @@
+#include <assert.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "monoclock.h"
+#include "warnp.h"
+
+#include "perftest.h"
+
+/**
+ * perftest_buffers(num_bytes, sizes, num_sizes, num_warm_up, init_func, func,
+ *     cookie):
+ * Time using ${func} to process ${num_bytes} bytes in blocks of ${sizes}.
+ * Before timing any block sizes, process ${num_warm_up} blocks of the maximum
+ * size in ${sizes}.  Before timing each block size, call
+ * ${(init_func)(cookie, buffer, buflen).  Print the time and speed of
+ * processing each buffer size.  ${func} will be invoked with (cookie, buffer,
+ * buflen, num_buffers), where ${buffer} is large enough to hold the maximum
+ * buffer size.
+ */
+int
+perftest_buffers(size_t num_bytes, const size_t * sizes, size_t num_sizes,
+    size_t num_warm_up,
+    int init_func(void * cookie, uint8_t * buf, size_t buflen),
+    int func(void * cookie, uint8_t * buf, size_t buflen, size_t num_buffers),
+    void * cookie)
+{
+	uint8_t * buf;
+	struct timeval begin, end;
+	double * delta_s;
+	double speed;
+	size_t i;
+	size_t buflen;
+	size_t num_buffers;
+	size_t max_buflen = 0;
+
+	/* Find the maximum buffer size. */
+	for (i = 0; i < num_sizes; i++) {
+		if (max_buflen < sizes[i])
+			max_buflen = sizes[i];
+	}
+
+	/* Allocate buffers. */
+	if ((buf = malloc(max_buflen)) == NULL) {
+		warnp("malloc");
+		goto err0;
+	}
+	if ((delta_s = malloc(max_buflen * sizeof(double))) == NULL) {
+		warnp("malloc");
+		goto err1;
+	}
+
+	/* Warm up */
+	if (init_func(cookie, buf, max_buflen))
+		goto err2;
+	if (func(cookie, buf, max_buflen, num_warm_up))
+		goto err2;
+
+	/* Run operations. */
+	for (i = 0; i < num_sizes; i++) {
+		/* Configure and sanity checks. */
+		buflen = sizes[i];
+		assert(buflen > 0);
+		num_buffers = num_bytes / buflen;
+		assert((num_bytes % buflen) == 0);
```

Comment:
I would be inclined to relax this -- it would allow #309 to have more buffer sizes (I would probably go up to 1024).

Reply:
If we do that, then different buffer sizes wouldn't be comparing the same amount of data.  (Also, at the moment the speed is calculated based on `num_bytes`, but I could change that to `buflen * num_buffers` if necessary)

For #309, I was thinking that we should tweak the number of bytes so that it's a multiple of any buffer size we want to check.  If that means that we change a bufsize of 10000 to 8192 or 9728, that sounds fine to me?

Reply:
BTW, that was an empirically-added `assert()`.  Namely, I added some more buffer sizes to one of the tests, didn't realize that they weren't an integer division, and spent 15 minutes trying to figure out why those buffer sizes were faster than the other ones.

(I think it was in crypto_aesctr, and I was testing the 1028-byte buffers we get from spiped.)

Reply:
Yes, I meant to change the speed calculation to use `buflen * num_buffers`.  And this should result in the same MB/s speeds, even if a (slightly!) different number of bytes is processed.